Короновирус застал мир в расплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещяют кафе и торговые центры.  Зато стали больше времени для книг . Это заметили стартаперы – и бросились создавать приложение для тех, кто любит читать.

Моя компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

Моя первая задача проанализировать базу данных.
В ней – информация о книгах, издателях авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта. 

### загрузим необходимые библиотеки

In [1]:
!pip install sqlalchemy

In [2]:
pip install psycopg2-binary

Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

# импортируем библиотеки
from sqlalchemy import create_engine 
import pandas as pd
from sqlalchemy import text, create_engine

In [4]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
con=engine.connect()

Всего из базы взято 5 таблиц. Откроем и просмотрим первые строки каждой таблицы для последующей рабльы с данными.

Посмотрим на таблицу books

In [5]:
query = '''SELECT * FROM books LIMIT 10 '''
books = pd.io.sql.read_sql(sql=text(query), con = con)
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

Содержит данные о книгах:
book_id — идентификатор книги;
author_id — идентификатор автора;
title — название книги;
num_pages — количество страниц;
publisher_id — идентификатор издателя.

Посмотрим на таблицу authors

In [6]:
query = '''SELECT * FROM authors LIMIT 10 '''
authors = pd.io.sql.read_sql(sql=text(query), con = con)
authors

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

Содержит данные об авторах:
author_id — идентификатор автора;
author — имя автора.

Посмотрим на таблицу publishers

In [7]:
query = '''SELECT * FROM publishers LIMIT 10 '''
publishers = pd.io.sql.read_sql(sql=text(query), con = con)
publishers

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

Содержит данные об издательствах:
publisher_id — идентификатор издательства;
publisher — название издательства;

Посмотрим на таблицу ratings

In [8]:
query = '''SELECT * FROM ratings LIMIT 10 '''
ratings = pd.io.sql.read_sql(sql=text(query), con = con)
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

Содержит данные о пользовательских оценках книг:
rating_id — идентификатор оценки;

book_id — идентификатор книги;
username — имя пользователя, оставившего оценку;
rating — оценка книги.




Посмотрим на таблицу reviews

In [9]:
query = '''SELECT * FROM reviews LIMIT 10 '''
reviews = pd.io.sql.read_sql(sql=text(query), con = con)
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

Содержит данные о пользовательских обзорах на книги:
review_id — идентификатор обзора;
book_id — идентификатор книги;
username — имя пользователя, написавшего обзор;
text — текст обзора.


таблица books соединяется по ключевому столбцу book_id с таблицами ratings и reviews. \
таблица books соединяется по ключевому столбцу author_id с таблицой authors. \
таблица books соединяется по ключевому столбцу publisher_id с таблицой publishers .

Мне как джуниор аналитику доверили сделать следующие запросы:

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [10]:
query = ''' SELECT 
                 COUNT(book_id)
            FROM 
                books 
            WHERE 
                publication_date > '2000-01-01'
            '''
books = pd.io.sql.read_sql(sql=text(query), con = con)
books

count
0    819

Всего вышло после 1 января 2000 года 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [11]:
query =     '''
            SELECT
               b.title,
               b.book_id,
               COUNT(DISTINCT re.review_id),
               AVG(ra.rating) AS avg_rating
            FROM 
                books AS b
            LEFT JOIN 
                reviews AS re ON b.book_id  = re.book_id
            JOIN
                ratings AS ra ON b.book_id  = ra.book_id 
            GROUP BY
                b.title,
                b.book_id
           
            '''

books = pd.io.sql.read_sql(sql=text(query), con = con)
books

title  book_id  count  \
0                                         'Salem's Lot        1      2   
1                   1 000 Places to See Before You Die        2      1   
2    13 Little Blue Envelopes (Little Blue Envelope...        3      3   
3    1491: New Revelations of the Americas Before C...        4      2   
4                                                 1776        5      4   
..                                                 ...      ...    ...   
995           Wyrd Sisters (Discworld  #6; Witches #2)      996      3   
996                        Xenocide (Ender's Saga  #3)      997      3   
997                                    Year of Wonders      998      4   
998                        You Suck (A Love Story  #2)      999      2   
999  Zen and the Art of Motorcycle Maintenance: An ...     1000      4   

     avg_rating  
0      3.666667  
1      2.500000  
2      4.666667  
3      4.500000  
4      4.000000  
..          ...  
995    3.666667  
996    3.400000  
997    3.200000  
998    4.500000  
999    3.833333  

[1000 rows x 4 columns]

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [12]:
query =     '''
            SELECT
               publisher,
               COUNT(book_id) AS books
              
            FROM 
                books AS b
            JOIN 
                publishers AS p ON b.publisher_id  = p.publisher_id
            WHERE
                num_pages >50
            GROUP BY
                publisher
          ORDER BY
              books DESC
        LIMIT 1
            '''

books = pd.io.sql.read_sql(sql=text(query), con = con)
books

publisher  books
0  Penguin Books     42

наибольшее количество книг выпустило издание – Penguin Books 42 книги.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [13]:
query = '''
    SELECT 
        author,
        AVG(rating) AS AVG_rating
    FROM (SELECT a.author,b.book_id,
           COUNT(rating_id) AS quantity_rating,
           AVG(rating) AS rating
            
           FROM books AS b   
           JOIN ratings AS ra ON b.book_id  = ra.book_id 
           JOIN authors AS a ON b.author_id  = a.author_id
           
           GROUP BY
              b.book_id,a.author
           HAVING
              COUNT(rating_id) >= 50) AS autors
     GROUP BY 
         author 
     ORDER BY
         AVG_rating DESC
     LIMIT 1

              '''


books = pd.io.sql.read_sql(sql=text(query), con = con)
books

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

самым популярным автором с более 50 оценками – J.K. Rowling/Mary GrandPré со средней оценкой 4.3 балла

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [14]:
query = '''
        SELECT
            AVG(quantity_review)
        FROM
            (SELECT  
               COUNT(review_id) AS quantity_review
            FROM 
                reviews
            WHERE
               username IN 
               
            (SELECT  
                username
            FROM 
                ratings AS ra   
             GROUP BY
                 username
             HAVING
                COUNT(rating) > 48)
                
             GROUP BY
                 username)
                 
                 AS review
                
            
        '''
books = pd.io.sql.read_sql(sql=text(query), con = con)
books

avg
0  24.0

пользователи которые поставили больше 48 оценок имеют в среднем 24 обзора 

Исходя из поставленных задач можно сказать что в нашем сервисе для чтения книг по подписке:
 -  после 1 января 2000 г. всего вышло  819 книг.
 - наибольшее количество книг выпустило издание – Penguin Books 42 книги.
 - самым популярным автором с более 50 оценками – J.K. Rowling/Mary GrandPré со средней оценкой 4.3 балла
 - пользователи которые поставили больше 48 оценок имеют в среднем 24 обзора 